components/scheduler/executor_executors.go

```diff
@@ -0,0 +1,493 @@
+
+	// Update SchedulerInfo metadata.
+	err = hsm.MachineTransition(schedulerNode, func(s Scheduler) (hsm.TransitionOutput, error) {
+		s.Info = scheduler.Info
+		return hsm.TransitionOutput{}, nil
+	})
+	if err != nil {
+		return err
+	}
+
+	// Find the earliest possible time a BufferedStart can be retried, if any, based
+	// on BackoffTime.
+	var retryAt time.Time
+	for _, start := range executor.BufferedStarts {
+		backoffTime := start.BackoffTime.AsTime()
+
+		// Zero-value protobuf timestamps deserialize to the UNIX epoch, skip if the
+		// backoffTime looks invalid/unset.
+		if backoffTime.Before(start.ActualTime.AsTime()) {
+			continue
+		}
+
+		if retryAt.IsZero() || backoffTime.Before(retryAt) {
+			retryAt = backoffTime
+		}
+	}
+
+	if !retryAt.IsZero() {
+		// When retryAt is set, reschedule.
+		return hsm.MachineTransition(node, func(e Executor) (hsm.TransitionOutput, error) {
+			e.ExecutorInternal.BufferedStarts = executor.BufferedStarts
+			return TransitionExecute.Apply(e, EventExecute{
+				Node:     node,
+				Deadline: retryAt,
+			})
+		})
+	}
+
+	// No more buffered starts, or remaining starts are waiting for a workflow to be
+	// closed. We can transition to waiting.
+	return hsm.MachineTransition(node, func(e Executor) (hsm.TransitionOutput, error) {
+		e.ExecutorInternal.BufferedStarts = executor.BufferedStarts
+		return TransitionWait.Apply(e, EventWait{
+			Node: node,
+		})
+	})
+}
+
+// drainBuffer uses ProcessBuffer to resolve the Executor's remaining buffered
+// starts, and then carries out the returned action.
+func (e executorTaskExecutor) drainBuffer(
+	logger log.Logger,
+	env hsm.Environment,
+	executor Executor,
+	scheduler Scheduler,
+	request *workflowpb.NewWorkflowExecutionInfo,
+) {
+	metricsWithTag := e.MetricsHandler.WithTags(
+		metrics.StringTag(metrics.ScheduleActionTypeTag, metrics.ScheduleActionStartWorkflow))
+	isRunning := len(scheduler.Info.RunningWorkflows) > 0
+
+	// Resolve overlap policies and prepare next workflows to start.
+	action := scheduler1.ProcessBuffer(executor.BufferedStarts, isRunning, scheduler.resolveOverlapPolicy)
+	newBuffer := action.NewBuffer
+
+	// Combine all available starts.
+	allStarts := action.OverlappingStarts
+	if action.NonOverlappingStart != nil {
+		allStarts = append(allStarts, action.NonOverlappingStart)
+	}
+
+	// Start workflows.
+	for _, start := range allStarts {
+		// Ensure we can take more actions. Manual actions are always allowed.
+		if !start.Manual && !scheduler.useScheduledAction(true) {
+			// Drop buffered automated actions while paused.
+			continue
+		}
+
+		if env.Now().Before(start.BackoffTime.AsTime()) {
+			// BufferedStart is still backing off, push it back into the buffer.
+			newBuffer = append(newBuffer, start)
+			continue
+		}
+
+		if env.Now().After(e.startWorkflowDeadline(scheduler, start)) {
+			// Drop expired starts.
+			continue
+		}
+
+		result, err := e.startWorkflow(env, scheduler, start, request)
```

Comment:
You shouldn't do any IO while holding the lock, you should do this in an immediate task, not a timer task (since the HSM framework can only execute timer tasks under a lock).